app/controllers/api/v1/meetings_controller.rb

```diff
@@ -81,25 +83,32 @@ def find_room
         @room = Room.find_by!(friendly_id: params[:friendly_id])
       end
 
-      def authorized_as_viewer?(access_code:)
-        return true if access_code.blank?
+      def authorized_as_viewer?(viewer_code:)
+        return true if viewer_code.blank? && params[:access_code].blank?
 
-        access_code == params[:access_code]
+        viewer_code == params[:access_code]
       end
 
-      def authorized_as_moderator?(access_code:)
-        (params[:access_code].present? && access_code == params[:access_code]) ||
-          @room.anyone_joins_as_moderator? ||
+      # Five scenarios where a user is authorized to join a BBB meeting as a moderator
+      # The user joins
+      # 1. its own room
+      # 2. a shared room
+      # 3. a room that requires a moderator access code and the access code input is correct
+      # 4. a room that has the AnyoneJoinAsModerator setting toggled on and does not require an access code
+      # 5. a room that has the AnyoneJoinAsModerator setting toggled on and requires a moderator or a viewer access code
+      #    and the access code input correspond to either the moderator or the viewer access code
+      def authorized_as_moderator?(mod_code:, viewer_code:, anyone_join_as_mod:)
+        @room.user_id == current_user&.id ||
           current_user&.shared_rooms&.include?(@room) ||
-          @room.user_id == current_user&.id
+          (params[:access_code].present? && mod_code == params[:access_code]) ||
+          (anyone_join_as_mod && mod_code.blank? && viewer_code.blank?) ||
+          (anyone_join_as_mod && (mod_code == params[:access_code] || viewer_code == params[:access_code]))
```

Comment:
After making e2e tests,
The following scenario failed:
```
[When anyone join as mod is ON]
Access code only:
	+ If entered join as mod [PASSED]
	+ If not entered or entered incorrectly forbid [FAILED, was able to bypass the access code check by an empty code]
```
I was able to bypass the access code check when giving an empty value when the access code is present but the moderator code is blank.
That is due to this line checking values directly without asserting the presence of codes.
This creates a security vulnerability,

Reply:
In addition, the `anyone joins as moderator` setting will always be treated as ON no matter what,
After isolating the cause, it's because of using `anyone_join_as_mod` as if it was a Boolean value,
The rooms meeting options value attribute contains strings not Booleans,
Boolean settings will have a stringified Boolean value like `'true'` but the type will be a string,
A string in ruby is truthy meaning that no matter what value is held it will be casted to a Boolean `true`,

Reply:
Ah, right.
Thank you for your time @KH-Amir-TN

Reply:
> 

Not sure of this, isn't it expected?
If Moderator Access Code is generated, and you don't provide the access code when joining, it is expected that you will join as a regular Viewer.